WalletWasabi.Tests/UnitTests/WabiSabi/Models/MultipartyTransaction.cs

```diff
@@ -0,0 +1,273 @@
+			Assert.Equal(WabiSabiProtocolErrorCode.WrongCoinjoinSignature, Assert.Throws<WabiSabiProtocolException>(() => noFeeTx.AddWitness(1, alice2Tx.Inputs[0].WitScript)).ErrorCode);
+
+			// Add Alice 1's signature
+			var alice1Sig = noFeeTx.AddWitness(0, alice1Tx.Inputs[0].WitScript);
+			Assert.False(alice1Sig.IsFullySigned);
+
+			// Witness can only be accepted once per input
+			Assert.Equal(WabiSabiProtocolErrorCode.WitnessAlreadyProvided, Assert.Throws<WabiSabiProtocolException>(() => alice1Sig.AddWitness(0, alice1Tx.Inputs[0].WitScript)).ErrorCode);
+		}
+
+		[Fact]
+		public void FeeRateValidation()
+		{
+			var feeRate = new FeeRate(new Money(1000L));
+
+			using Key key1 = new();
+			using Key key2 = new();
+
+			var alice1 = WabiSabiFactory.CreateAlice(key: key1);
+			var alice2 = WabiSabiFactory.CreateAlice(key: key2);
+
+			var state = new Construction(DefaultParameters with { FeeRate = feeRate })
+				.AddInput(alice1.Coins.First())
+				.AddInput(alice2.Coins.First());
+
+			var bob1 = new TxOut(Money.Coins(1), alice1.Coins.First().ScriptPubKey);
+			var withOutput = state.AddOutput(bob1);
+
+			Assert.Equal(2, withOutput.Inputs.Count());
+			Assert.Single(withOutput.Outputs);
+
+			var bob2 = new TxOut(Money.Coins(1), alice2.Coins.First().ScriptPubKey);
+			Assert.Equal(WabiSabiProtocolErrorCode.InsufficientFees, Assert.Throws<WabiSabiProtocolException>(() => withOutput.AddOutput(bob2).Finalize()).ErrorCode);
+
+			bob2 = new TxOut(Money.Coins(0.9999m), alice2.Coins.First().ScriptPubKey);
+			var generousFeeTx = withOutput.AddOutput(bob2).Finalize();
+
+			var tx = generousFeeTx.CreateUnsignedTransaction();
+			Assert.Equal(2, tx.Inputs.Count);
+			Assert.Equal(2, tx.Outputs.Count);
+			Assert.Contains(alice1.Coins.First().Outpoint, tx.Inputs.Select(x => x.PrevOut));
+			Assert.Contains(alice2.Coins.First().Outpoint, tx.Inputs.Select(x => x.PrevOut));
+			Assert.Contains(alice1.Coins.First().ScriptPubKey, tx.Outputs.Select(x => x.ScriptPubKey));
+			Assert.Contains(alice2.Coins.First().ScriptPubKey, tx.Outputs.Select(x => x.ScriptPubKey));
+
+			var alice1Tx = tx.Clone();
+			alice1Tx.Sign(key1.GetBitcoinSecret(Network.Main), alice1.Coins.First());
+
+			var alice1Sig = generousFeeTx.AddWitness(0, alice1Tx.Inputs[0].WitScript);
+			Assert.False(alice1Sig.IsFullySigned);
+			Assert.Equal(alice1Tx.ToString(), alice1Sig.CreateTransaction().ToString());
+
+			var alice2Tx = tx.Clone();
+			alice2Tx.Sign(key2.GetBitcoinSecret(Network.Main), alice2.Coins.First());
+
+			var alice2Sig = alice1Sig.AddWitness(1, alice2Tx.Inputs[1].WitScript);
+			Assert.True(alice2Sig.IsFullySigned);
+
+			var signed = alice2Sig.CreateTransaction();
+			Assert.NotEqual(alice1Tx.ToString(), signed.ToString());
+			Assert.NotEqual(alice2Tx.ToString(), signed.ToString());
+			Assert.True(signed.Inputs.All(x => x.HasWitScript()));
```

Comment:
This is exactly the same but I think the xUnit failure report would be more useful:

```suggestion
			Assert.All(signed.Inputs, x => Assert.True(x.HasWitScript()));
```